File: JS_Submissions/sliding_window/exercise1.js
/*
Maximum Average Subarray I

Solution
You are given an integer array nums consisting of n elements, and an integer k.

Find a contiguous subarray whose length is equal to k that has the maximum average value and return this value. Any answer with a calculation error less than 10-5 will be accepted.

 

Example 1:

Input: nums = [1,12,-5,-6,50,3], k = 4
Output: 12.75000
Explanation: Maximum average is (12 - 5 - 6 + 50) / 4 = 51 / 4 = 12.75
Example 2:

Input: nums = [5], k = 1
Output: 5.00000
 

Constraints:

n == nums.length
1 <= k <= n <= 105
-104 <= nums[i] <= 104

*/


/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findMaxAverage = function(nums, k) {
    let ans = 0, currSum = 0;
    
    for (let i = 0; i < k; i++) {
        currSum += nums[i];
    }
    
    ans = currSum / k;
    
    for(let i = k; i < nums.length; i++) {
        currSum += nums[i] - nums[i - k];
        ans = Math.max(ans, currSum / k);
    }
    
    return ans;
};

// both loops cover the entire length of nums so this algorithm has a runtime of O(n),
// where n is the length of nums, and O(1) space.

File: Rust_Submissions/prefix_sum/exercise1.rs
/*

Running Sum of 1d Array

Solution
Given an array nums. We define a running sum of an array as runningSum[i] = sum(nums[0]…nums[i]).

Return the running sum of nums.

 

Example 1:

Input: nums = [1,2,3,4]
Output: [1,3,6,10]
Explanation: Running sum is obtained as follows: [1, 1+2, 1+2+3, 1+2+3+4].
Example 2:

Input: nums = [1,1,1,1,1]
Output: [1,2,3,4,5]
Explanation: Running sum is obtained as follows: [1, 1+1, 1+1+1, 1+1+1+1, 1+1+1+1+1].
Example 3:

Input: nums = [3,1,2,10,1]
Output: [3,4,6,16,17]
 

Constraints:

1 <= nums.length <= 1000
-10^6 <= nums[i] <= 10^6

*/

fn running_sum(nums: Vec<i32>) -> Vec<i32> {
    let len = nums.len();
    let mut start = 1 as usize;
    let mut prefix = vec![0; len];
    prefix[0] = nums[0];
        
    for i in start..len {
        prefix[i] = nums[i] + prefix[i - 1];
    }
        
    prefix
}